### DISCUSSION TITLE
Add support for express/connect middlewares

### DISCUSSION THREAD

**Discussion by rafaell-lycan** [honojs/hono#1520] [Ideas]
Hello there 👋 

I'd like to ask about the possibility of creating a way to use express/connect syntax-style middleware with a plug-and-play solution.

Fastify does it using [middie](https://github.com/fastify/middie) and it just works without sacrificing performance.

I'm migrating one API at a time starting the internal ones that don't require too much stuff/middleware.

**Why:**
I'm migrating a few Express apps that have some custom middleware in between and using Express/Connect signatures.

Some example:
- Metrics & Tracing (using StatsD + OTL)
- Auth (Custom: uses headers and fallback to cookie)
- Rate Limit (Custom: uses headers and fallback to cookie)
- HTTP Client (Custom: uses Axios with headers support to keep everything traceable)
- Logger (with Pino)

Hopefully, this won't be something useful just for me, as [multer](https://www.npmjs.com/package/multer) and [helmet](https://www.npmjs.com/package/helmet) would be an amazing addition to it.

**Comment by yusukebe** [comment]
The Express/connect middleware is based on Node.js incoming/outgoing messages. To be able to use them, the Node.js Adapter is required. Therefore, it is difficult to use many of them in an edge environment. Also, Hono will be slow with the Node.js Adapter, so it is impossible to introduce them while maintaining performance.

**Comment by rafaell-lycan** [comment]
@yusukebe the idea was to use it in conjunction with the Node.js Adapter as well, as it only simplifies when migration from an Express project to Hono.

That's why this would be a 3rd-party middleware, meaning the Node.js Adapter would be mandatory/required.

Again, this is my current use case to avoid rewriting an Hono version of each one to migrate a dozen APIs, instead, it could bring some simplicity like [middie](https://github.com/fastify/middie) does for Fastify.

**Reply by flipvh** [reply]
Hi Rafaell, hope you are doing well! I got here looking for honojs with pino actually. Do you have experience with this combo already? :D Perhaps we can collaborate on middleware if you are still into honojs. I think I have the same predicament actually.

**Reply by rafaell-lycan** [reply]
Hey @flipvh 👋 

I've created a custom middleware based on [Hono's Logger middleware](https://github.com/honojs/hono/blob/main/src/middleware/logger/index.ts) but using Pino instead.

You can also [plug your pino fn](https://hono.dev/middleware/builtin/logger#example) directly on the `logger`, but as we have other custom middleware that we need to collect data as well, we decided to create something custom.

I've spent much time adapting our metrics/tracing libs to Hono, but after a few days, everything worked.

The only middleware I might be able to add here will be the `rate-limit` based on `express-rate-limit` with a Redis adapter.

Still need some refactoring first but that's the way, unless someone publishes something first.

**Reply by flipvh** [reply]
thanks for sharing! I ended up working on a copy-paste of the hono logger - after finding out it was just one small file! - and tuning it to my liking: https://github.com/cellajs/cella/blob/main/backend/src/middlewares/logger/logger.ts

metrics/tracing is still a work in progress. You also use Sentry? And yes, [we also use express-rate-limiter but it needs a refactor still.](https://github.com/cellajs/cella/blob/main/backend/src/middlewares/rate-limiter/index.ts)

For a route-specific guard middleware we added an abstraction layer on top of zod-openapi, but this is [also a wip](https://github.com/honojs/middleware/issues/417). Let me know if you see possibilities to collaborate! all the best Flip

**Reply by rafaell-lycan** [reply]
@flipvh for general metrics/tracing, I'd recommend you take a look at [opentelemetry-instrumentation-express](https://github.com/open-telemetry/opentelemetry-js-contrib/tree/main/plugins/node/opentelemetry-instrumentation-express) or any other plugin they have. Also for simple metrics, you can create your collector using something like [hot-shots](https://github.com/brightcove/hot-shots).

TL;DR we migrated from [Datadog JS tracing](https://github.com/DataDog/dd-trace-js) in favor of something generic as OTLP.
